client/client.go

```diff
@@ -28,120 +28,114 @@ var (
+		logger: logger,
 	}
-
-	ctx, _ := context.WithTimeout(context.Background(), g.timeout)
-	_, err := g.client.SendEvent(ctx, req)
-	return err
-}
-
-// Send sends an event to another server via grpc using the client's configured topic
-func (g *GRPCClient) Send(name string, props map[string]string) error {
-	l := g.logger.WithFields(log.Fields{
-		"operation": "send",
-		"event":     name,
-	})
-	l.Debug("sending event")
-	err := g.sendEvent(name, g.topic, props)
+	if configPrefix != "" && !strings.HasSuffix(configPrefix, ".") {
+		configPrefix = strings.Join([]string{configPrefix, "."}, "")
+	}
+	err := g.configure(configPrefix, client)
 	if err != nil {
-		l.WithError(err).Error("send event failed")
+		g.logger.WithError(err).Error("failed to configure client")
+		return nil, err
 	}
-	l.Debug("successfully sended event")
-	return err
+	if g.async {
+		for i := 0; i < g.numSendEventsRoutines; i++ {
+			go g.sendEventsRoutine()
+		}
+	}
+	return g, nil
 }
 
-// SendToTopic sends an event to another server via grpc using an explicit topic
-func (g *GRPCClient) SendToTopic(name, topic string, props map[string]string) error {
-	l := g.logger.WithFields(log.Fields{
-		"operation": "sendToTopic",
-		"event":     name,
-		"topic":     topic,
-	})
-	l.Debug("sending event")
-	err := g.sendEvent(name, topic, props)
-	if err != nil {
-		l.WithError(err).Error("send event failed")
+func (g *GRPCClient) configure(configPrefix string, client pb.GRPCForwarderClient) error {
+	if err := g.configureSend(configPrefix); err != nil {
+		return err
 	}
-	l.Debug("successfully sended event")
-	return err
+	if err := g.configureGRPC(configPrefix, client); err != nil {
+		return err
+	}
+	return nil
 }
 
-// MetricsReporterInterceptor will report metrics from client
-func (g *GRPCClient) MetricsReporterInterceptor(
-	ctx context.Context,
-	method string,
-	req interface{},
-	reply interface{},
-	cc *grpc.ClientConn,
-	invoker grpc.UnaryInvoker,
-	opts ...grpc.CallOption,
-) error {
-	l := g.logger.WithFields(log.Fields{
-		"method": method,
-	})
+func (g *GRPCClient) configureSend(configPrefix string) error {
+	asyncConf := fmt.Sprintf("%sclient.send.async", configPrefix)
+	g.config.SetDefault(asyncConf, false)
+	g.async = g.config.GetBool(asyncConf)
 
-	startTime := time.Now()
-	event := req.(*pb.Event)
+	asyncRetriesConf := fmt.Sprintf("%sclient.send.asyncRetries", configPrefix)
+	g.config.SetDefault(asyncRetriesConf, false)
+	g.asyncRetries = g.config.GetInt(asyncRetriesConf)
 
-	defer func() {
-		timeUsed := float64(time.Since(startTime).Nanoseconds() / int64(time.Millisecond))
-		metrics.ClientRequestsResponseTime.WithLabelValues(hostname, method, event.GetTopic()).Observe(timeUsed)
-		l.WithFields(log.Fields{
-			"timeUsed": timeUsed,
-			"reply":    reply.(*pb.Response),
-		}).Debug("request processed")
-	}()
+	asyncRetryIntervalConf := fmt.Sprintf("%sclient.send.asyncRetryInterval", configPrefix)
+	g.config.SetDefault(asyncRetryIntervalConf, false)
```

Comment:
probably a bad copy and paste here, interval should be a duration ;)

Reply:
woops